cmd/clusterctl/cmd/init.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+)
+
+type initOptions struct {
+	kubeconfig              string
+	coreProvider            string
+	bootstrapProviders      []string
+	infrastructureProviders []string
+	targetNamespace         string
+	watchingNamespace       string
+	force                   bool
+}
+
+var io = &initOptions{}
+
+var initCmd = &cobra.Command{
+	Use:   "init",
+	Short: "Initialize a management cluster for Cluster API",
+	Long: LongDesc(`
+		Initialize a management cluster for Cluster API by installing cluster API core components,
+		the kubeadm bootstrap provider, and the selected bootstrap and infrastructure providers.
+
+		The management cluster should be an existing Kubernetes cluster, and the identity used
+		for accessing the cluster must have enough privileges for installing cluster API providers.
+
+		Use 'clusterctl config providers' to get the list of available providers; if necessary, edit
+		the .clusterctl file to add new provider configurations or to customize existing ones.
+		
+		Some providers require environment variable to be set before running clusterctl init; please
+		refer to the provider documentation or use 'clusterctl config provider [name]' to get the
+		list of required variables.
+
+		See https://cluster-api.sigs.k8s.io/ for more details.`),
+
+	Example: Examples(`
+		# Initialize a management cluster, the cluster currently pointed by kubectl, by installing the
+		# AWS infrastructure provider. Please note that this command, when executed on an empty management cluster, 
+ 		# automatically triggers the installation of the cluster-api core provider.
+		clusterctl init --infrastructure=aws
+
+		# Initialize a management cluster with a specific version of the AWS infrastructure provider.
+		clusterctl init --infrastructure=aws:v0.4.1
+
+		# Initialize a management cluster, the cluster defined in the "foo.yaml" kubeconfig file, by 
+		# installing the AWS infrastructure provider.
+		clusterctl init --kubeconfig=foo.yaml  --infrastructure=aws
+
+		# Initialize a management cluster, by installing both the AWS and the vSphere infrastructure provider
+		clusterctl init --infrastructure=aws,vsphere
+
+		# Initialize a management cluster by installing the provider's components' in the "foo" namespace.
+		clusterctl init --infrastructure aws --target-namespace foo
```

Comment:
In general I'm curious about the value of a namespace override here. Could we assume the namespace will be `<provider>-system` (e.g. `capa-system`, `capv-system`, etc) and not worry about this?

Reply:
Each provider comes with a default namespace e.g. `capv-system` for CAPV.
The `--target-namespace` flag should be used when the want to override this value, and in that it is up to the user to choose the value to use here.
AFAIK, there are users deploying all the providers into a single namespace, but I don't know the naming convention used for such object

Reply:
Would it be possible to assume the well-known namespace per provider (capv-system, capa-system, etc) and add this feature later if users request it? I feel it adds complexity to the UX and I think we should exclude it unless we're really confident about the use-case. Of the users deploying all the providers into a single namespace what was the rational?

Reply:
@andrewsykim I will be more than happy in getting rid of the namespace management, but until there is no clear decision about https://github.com/kubernetes-sigs/cluster-api/issues/1631, I can't rule out that a user will be required to install more instance of the same provider, each one in different namespaces.
Once this is addressed, I can eventually rework this.

Reply:
I propose we proceed - this is just a scaffolding. We will have room to iterate.